lmdeploy/serve/async_engine.py

```diff
@@ -471,9 +471,14 @@ async def generate(
         if do_preprocess:
             prompt = self.chat_template.messages2prompt(prompt, sequence_start)
         input_ids = self.tokenizer.encode(prompt, add_bos=sequence_start)
+        if gen_config.max_new_tokens is None:
+            # for interactive endpoint, will try maximum possible token num
+            gen_config.max_new_tokens = max(
+                128, self.session_len - self.id2step[str(session_id)] -
+                len(input_ids))
         finish_reason = None
         if self.id2step[str(session_id)] + len(
-                input_ids) + gen_config.max_new_tokens >= self.session_len:
+                input_ids) + gen_config.max_new_tokens > self.session_len:
```

Comment:
If max_tokens is left to the application to set, it may result in empty returns due to mistakes. Instead, is it possible to calculate the remaining number of tokens automatically based on the supported max_tokens of the model minus the input tokens? If the automatically calculated remaining number is less than 0, an exception should be returned, indicating that the maximum context limit of the model has been exceeded. Otherwise, the smaller of the remaining token count and the specified max_tokens should be used.

Reply:
> it was computed automatically when user input `max_tokens= None`. Please check the line 474 of this PR.

Thank you, I noticed it. Due to the uncertainty of the length of user input in the application, max_tokens may not be set in some scenarios, and this parameter needs to be carefully calculated in very few scenarios. Anyway, thank you for your answer. For now, I can set max_tokens to 0 to achieve the same effect as not setting it.